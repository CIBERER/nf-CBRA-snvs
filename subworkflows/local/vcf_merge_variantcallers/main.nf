include { BCFTOOLS_MERGE } from '../../../modules/nf-core/bcftools/merge/main'
include { TABIX_TABIX } from '../../../modules/nf-core/tabix/tabix/main'
include { TABIX_TABIX as TABIX_TABIX_FINAL_VCF } from '../../../modules/nf-core/tabix/tabix/main'
include { BCFTOOLS_QUERY_STATS } from '../../../modules/local/bcftools_query_stats/main'
include { CONSENSUS_GENOTYPE } from '../../../modules/local/consensus_genotype/main'
include { GET_VCF_CALLERS_INFO } from '../../../modules/local/get_vcf_callers_info/main'
include { CREATE_SAMPLE_INFO } from '../../../modules/local/create_sample_info/main'



workflow VCF_MERGE_VARIANTCALLERS {

    take:
    ch_vcfs        // channel (mandatory): tuple val(meta), path(vcfs), path(tbis)
    ch_fasta        // channel (mandatory) : [ val(meta3), path(fasta) ]
    ch_fai          // channel (mandatory) : [ val(meta3), path(fai) ]
    ch_intervals    // channel (mandatory) : [ val(meta), path(bed) ]
    ch_assembly

    main:

    ch_versions = Channel.empty()

   ch_vcfs.map { item ->
        def meta = item[0]
        def files = item[1..-1].collect { file(it) }
        def vcfs = files.findAll { it.name.endsWith('.vcf.gz') }
        def tbis = files.findAll { it.name.endsWith('.vcf.gz.tbi') }
        [meta, vcfs, tbis]
    }
    .set { ch_for_bcftoolsmerge }

    ch_for_bcftoolsmerge.view()

    BCFTOOLS_MERGE (
        ch_for_bcftoolsmerge,
        ch_fasta,
        ch_fai,
        ch_intervals
    )

    TABIX_TABIX(
        BCFTOOLS_MERGE.out.vcf
    )

    BCFTOOLS_QUERY_STATS (
        BCFTOOLS_MERGE.out.vcf.join(TABIX_TABIX.out.tbi)
    )

    CONSENSUS_GENOTYPE (
        BCFTOOLS_QUERY_STATS.out.gt
    )

    GET_VCF_CALLERS_INFO (
        BCFTOOLS_QUERY_STATS.out.gt
    )

    ch_stats = CONSENSUS_GENOTYPE.out.consensus_gt.join(CONSENSUS_GENOTYPE.out.discordances).join(BCFTOOLS_QUERY_STATS.out.ad_mean).join(BCFTOOLS_QUERY_STATS.out.dp_mean).join(BCFTOOLS_QUERY_STATS.out.rd_mean).join(BCFTOOLS_QUERY_STATS.out.vd_mean).join(BCFTOOLS_QUERY_STATS.out.vaf).join(GET_VCF_CALLERS_INFO.out.sf_file).join(BCFTOOLS_QUERY_STATS.out.programs)

    CREATE_SAMPLE_INFO (
        BCFTOOLS_MERGE.out.vcf.join(TABIX_TABIX.out.tbi).join(ch_stats),
        ch_assembly
    )

    TABIX_TABIX_FINAL_VCF (
        CREATE_SAMPLE_INFO.out.final_vcf
    )

    vcf = CREATE_SAMPLE_INFO.out.final_vcf.join(TABIX_TABIX_FINAL_VCF.out.tbi).view()

    emit:
    vcf                                        // channel: [ [val(meta)], path(vcf), path(tbi)]
    versions = ch_versions                     // channel: [ versions.yml ]

}
