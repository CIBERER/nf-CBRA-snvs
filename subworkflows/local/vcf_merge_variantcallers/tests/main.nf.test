nextflow_workflow {

    name "Test Subworkflow VCF_MERGE_VARIANTCALLERS"
    script "../main.nf"
    workflow "VCF_MERGE_VARIANTCALLERS"

    tag "subworkflows"
    tag "subworkflows/vcf_merge_variantcallers"


   setup {

        run("BWA_INDEX") {
            script "../../../../modules/nf-core/bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("PICARD_CREATESEQUENCEDICTIONARY") {
            script "../../../../modules/nf-core/picard/createsequencedictionary/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("MAPPING") {
            script "../../../../subworkflows/local/mapping/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id : 'test_sample' ],
                    [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                ])
                input[1] =  Channel.fromList([
                            tuple([ id: 'test_sample'],
                            [])
                        ])
                input[2] = BWA_INDEX.out.index
                input[3] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[4] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[5] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true)
                input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi', checkIfExists: true)
                """
            }
        }

        run("GATK_VCF") {
            script "../../../../subworkflows/local/gatk_vcf/main.nf"
            process {
                """
                input[0] = MAPPING.out.bam
                input[1] =  Channel.fromList([
                        tuple([ id: 'test_sample'],
                        [])
                    ])
                input[2] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[3] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[4] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[5] = input[5] =  Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                input[6] = input[5] =  Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                """
            }
        }

        run("GATK4_COMPOSESTRTABLEFILE") {
            script "../../../../modules/nf-core/gatk4/composestrtablefile/main.nf"
            process {
                """
                input[0] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                input[1] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                input[2] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                """
            }
        }

        run("DRAGEN_VCF") {
            script "../../../../subworkflows/local/dragen_vcf/main.nf"
            process {
                 """
                input[0] = MAPPING.out.bam
                input[1] = Channel.fromList([
                        tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ])
                input[2] =  Channel.fromList([
                    tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[3] = Channel.fromList([
                    tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true))
                ])
                input[4] = GATK4_COMPOSESTRTABLEFILE.out.str_table
                input[5] =  Channel.fromList([
                        tuple([ id: 'test_sample'],
                        [])
                    ])                
                input[6] = Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                input[7] = Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])                
                """
            }
        }


    }


    test("vcf_merge_variantcallers - vcf ") {


        when {
            workflow {
                """
                input[0] = GATK_VCF.out.vcf.map { meta, vcf, tbi -> [meta + [program:"gatk"], vcf, tbi] }.concat(
                    DRAGEN_VCF.out.vcf.map { meta, vcf, tbi -> [meta + [program:"dragen"], vcf, tbi] }
                )
                input[1] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[2] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                input[3] =  Channel.fromList([
                        tuple([ id: 'test_sample'],
                        [])
                    ])    
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

}
