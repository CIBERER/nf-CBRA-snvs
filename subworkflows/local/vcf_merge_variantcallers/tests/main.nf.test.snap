{
    "vcf_merge_variantcallers - vcf ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.hg38.final.vcf.gz:md5,d65251375f8c7ce2f4fa7ac8019975f7",
                        "test_sample.hg38.final.vcf.gz.tbi:md5,fb90593543dc307fe9363e6e859ab6f2"
                    ]
                ],
                "1": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.hg38.final.vcf.gz:md5,d65251375f8c7ce2f4fa7ac8019975f7",
                        "test_sample.hg38.final.vcf.gz.tbi:md5,fb90593543dc307fe9363e6e859ab6f2"
                    ]
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-05-05T08:24:24.467654855"
    }
}